File: backend/internal/domain/models/specific_preference.go
package models

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// SpecificPreferenceType represents the type of preference
type SpecificPreferenceType string

const (
	SpecificPreferenceTypePositionCount SpecificPreferenceType = "position_count" // Requires a position and count
	SpecificPreferenceTypeStaffName     SpecificPreferenceType = "staff_name"     // Requires a specific staff member
)

// SpecificPreference represents a preference rule that can override or mix with other filters
// It allows setting preferences based on combinations of branch, doctor, and day of week
type SpecificPreference struct {
	ID          uuid.UUID              `json:"id" db:"id"`
	BranchID    *uuid.UUID             `json:"branch_id,omitempty" db:"branch_id"`     // NULL = any branch
	Branch      *Branch                `json:"branch,omitempty"`
	DoctorID    *uuid.UUID             `json:"doctor_id,omitempty" db:"doctor_id"`     // NULL = any doctor
	Doctor      *Doctor                `json:"doctor,omitempty"`
	DayOfWeek   *int                   `json:"day_of_week,omitempty" db:"day_of_week"` // NULL = any day, 0-6 (Sunday-Saturday)
	PreferenceType SpecificPreferenceType `json:"preference_type" db:"preference_type"`
	
	// For position_count type: position_id and staff_count are required
	PositionID  *uuid.UUID             `json:"position_id,omitempty" db:"position_id"`
	Position    *Position               `json:"position,omitempty"`
	StaffCount  *int                   `json:"staff_count,omitempty" db:"staff_count"`
	
	// For staff_name type: staff_id is required
	StaffID     *uuid.UUID             `json:"staff_id,omitempty" db:"staff_id"`
	Staff       *Staff                 `json:"staff,omitempty"`
	
	IsActive    bool                   `json:"is_active" db:"is_active"`
	CreatedAt   time.Time              `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time              `json:"updated_at" db:"updated_at"`
}

// Validate validates the SpecificPreference model
func (sp *SpecificPreference) Validate() error {
	if sp.PreferenceType == SpecificPreferenceTypePositionCount {
		if sp.PositionID == nil {
			return fmt.Errorf("position_id is required for position_count preference type")
		}
		if sp.StaffCount == nil || *sp.StaffCount < 1 {
			return fmt.Errorf("staff_count must be at least 1 for position_count preference type")
		}
	} else if sp.PreferenceType == SpecificPreferenceTypeStaffName {
		if sp.StaffID == nil {
			return fmt.Errorf("staff_id is required for staff_name preference type")
		}
	} else {
		return fmt.Errorf("invalid preference_type: %s", sp.PreferenceType)
	}
	return nil
}

// Matches checks if this preference matches the given branch, doctor, and day of week
func (sp *SpecificPreference) Matches(branchID *uuid.UUID, doctorID *uuid.UUID, dayOfWeek *int) bool {
	if !sp.IsActive {
		return false
	}
	
	// Check branch match (NULL = any branch)
	if sp.BranchID != nil {
		if branchID == nil || *sp.BranchID != *branchID {
			return false
		}
	}
	
	// Check doctor match (NULL = any doctor)
	if sp.DoctorID != nil {
		if doctorID == nil || *sp.DoctorID != *doctorID {
			return false
		}
	}
	
	// Check day of week match (NULL = any day)
	if sp.DayOfWeek != nil {
		if dayOfWeek == nil || *sp.DayOfWeek != *dayOfWeek {
			return false
		}
	}
	
	return true
}

File: frontend/src/lib/api/specific-preference.ts
import apiClient from './client';

export type SpecificPreferenceType = 'position_count' | 'staff_name';

export interface SpecificPreference {
  id: string;
  branch_id?: string;
  branch?: {
    id: string;
    name: string;
    code: string;
  };
  doctor_id?: string;
  doctor?: {
    id: string;
    name: string;
    code: string;
  };
  day_of_week?: number; // 0-6 (Sunday-Saturday), null for any day
  preference_type: SpecificPreferenceType;
  position_id?: string;
  position?: {
    id: string;
    name: string;
  };
  staff_count?: number;
  staff_id?: string;
  staff?: {
    id: string;
    name: string;
    nickname: string;
  };
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateSpecificPreferenceRequest {
  branch_id?: string | null; // null for "any branch"
  doctor_id?: string | null; // null for "any doctor"
  day_of_week?: number | null; // 0-6 or null for "any day"
  preference_type: SpecificPreferenceType;
  position_id?: string; // Required for position_count
  staff_count?: number; // Required for position_count
  staff_id?: string; // Required for staff_name
  is_active?: boolean;
}

export interface UpdateSpecificPreferenceRequest {
  branch_id?: string | null;
  doctor_id?: string | null;
  day_of_week?: number | null;
  preference_type?: SpecificPreferenceType;
  position_id?: string | null;
  staff_count?: number | null;
  staff_id?: string | null;
  is_active?: boolean;
}

export interface SpecificPreferenceFilters {
  branch_id?: string;
  doctor_id?: string;
  day_of_week?: number;
  is_active?: boolean;
}

export const specificPreferenceApi = {
  list: async (filters?: SpecificPreferenceFilters): Promise<{ preferences: SpecificPreference[] }> => {
    const params = new URLSearchParams();
    if (filters?.branch_id) params.append('branch_id', filters.branch_id);
    if (filters?.doctor_id) params.append('doctor_id', filters.doctor_id);
    if (filters?.day_of_week !== undefined) params.append('day_of_week', filters.day_of_week.toString());
    if (filters?.is_active !== undefined) params.append('is_active', filters.is_active.toString());
    
    const queryString = params.toString();
    const url = `/specific-preferences${queryString ? `?${queryString}` : ''}`;
    const response = await apiClient.get(url);
    return response.data;
  },

  getById: async (id: string): Promise<{ preference: SpecificPreference }> => {
    const response = await apiClient.get(`/specific-preferences/${id}`);
    return response.data;
  },

  create: async (data: CreateSpecificPreferenceRequest): Promise<{ preference: SpecificPreference }> => {
    const response = await apiClient.post('/specific-preferences', data);
    return response.data;
  },

  update: async (id: string, data: UpdateSpecificPreferenceRequest): Promise<{ preference: SpecificPreference }> => {
    const response = await apiClient.put(`/specific-preferences/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/specific-preferences/${id}`);
  },
};
